File: app/graphql/mutations/register_user.rb
# frozen_string_literal: true

module Mutations
  class RegisterUser < Mutations::BaseMutation
    graphql_name "RegisterUser"

    argument :first_name, String, required: true
    argument :last_name, String, required: true
    argument :email, String, required: true
    argument :password, String, required: true

    field :user, Types::UserType, null: false

    def resolve(args)
      user = User.create!(args)

      # current_user needs to be set so authenticationToken can be returned
      context[:current_user] = user

      MutationResult.call(
        obj: { user: user },
        success: user.persisted?,
        errors: user.errors
      )
    rescue ActiveRecord::RecordInvalid => invalid
      GraphQL::ExecutionError.new(
        "Invalid Attributes for #{invalid.record.class.name}: " \
        "#{invalid.record.errors.full_messages.join(', ')}"
      )
    end
  end
end

File: app/javascript/mutations/registerUser.js
// app/javascript/mutations/registerUser.js

import gql from 'graphql-tag';

const mutation = gql`
  mutation registerUser(
    $firstName: String!,
    $lastName: String!,
    $email: String!,
    $password: String!,
  ) {
    registerUser(input: {
      firstName: $firstName,
      lastName: $lastName,
      email: $email,
      password: $password,
    }) {
      user {
        id
        firstName
        lastName
        email
        authenticationToken
      }
      success
      errors
    }
  }
`;

export default function({
  apollo,
  firstName,
  lastName,
  email,
  password,
}) {
  return apollo.mutate({
    mutation,
    variables: {
      firstName,
      lastName,
      email,
      password,
    },
  });
}